HW3 Считывание и подсчет слов в отдельных горутинах

r-i-besedin opened the pull request:
Слова считываются и подсчитываются из файлов в отдельных горутинах. Для добавления в общий индекс используется mutex

r-i-besedin committed:
Первый коммит по поиску релевантного файла

Файлы для индексации вводятся без ключа аргументов, например:
        ./main ./search ex1.txt ex2.txt ex3.txt
Поисковая фраза - строка с ключевыми словами, разделенными пробелами
Программа выведет файлы в порядке наилучшего соответствия (слова из поисковой фразы встретились в файле наибольшее количество раз).

r-i-besedin committed:
Подсчет количества слов вынесен в пакет с индексом

Теперь в главном пакете формируется слайс слов из файла, а подсчитываются они уже в пакете с индексом

r-i-besedin committed:
Считывание файлов в отдельных горутинах

Слова из файлов считываются и подсчитываются в отдельных горутинах, это должно дать хороший прирост производительности для индексации больших файлов

tariel-x commented on invertedindex/invertedindex.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func AttachWordsOccurencesToGlobalMap(source string, strcmap map[string]int){
+	gSafeIndexMap.Lock()
```
Мьютекс на всю map или не ускорит, или замедлит работу. Лучше попробовать сделать с помощью каналов. Либо сделать одну горутину, читающую из канала и добавляющую в индекс и несколько горутин, которые читают файлы и пишут в канал. Либо в нескольких горутинах строить отдельный индексы, а потом сливать в один.

r-i-besedin replied:
> Либо сделать одну горутину, читающую из канала и добавляющую в индекс и несколько горутин, которые читают файлы и пишут в канал.

Сделал так; еще проверил реализацию на мьютексе и на канале по бенчмарк тесту, в общем реализации с одной скоростью работают. Единственное, что с мьютексом мне кажется более просто и логично, потому что даже после инициализации в дальнейшем можно добавлять новые слова

tariel-x replied:
Вы на больших файлах проверяли?

У меня на https://github.com/polis-mail-ru-golang-1/examples/tree/master/t2/books получилось для 3-х разных реализций:

```
BenchmarkSingle-2              3         368816069 ns/op        45617024 B/op     438439 allocs/op
BenchmarkParallel-2            5         228130356 ns/op        50435379 B/op     461308 allocs/op
BenchmarkMutex-2               3         362978362 ns/op        45614426 B/op     438429 allocs/op
```

r-i-besedin committed:
Перестроена работа многопоточности на использование каналов

Файлы разбиваются и считаются вхождения слов в отдельных горутинах. Другая горутина добавляет карту посчитанных слов в общий индекс

tariel-x commented:
:+1:
